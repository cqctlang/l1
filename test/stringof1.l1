define mkzas(len){
	@local dispatch, str, map;
	str = mkstr(len);
	map = vector(mkrange(0,len));
	define dispatch args {
		@local arg, range, s, new, beg, end;
		arg = listref(args, 1);

		if(arg == "get"){
			range = listref(args, 2);
			beg = rangebeg(range);
			end = beg+rangelen(range);
			printf("get %x %x\n", beg, end);
			return substr(str, beg, end);
		}

		if(arg == "put"){
			range = listref(args, 2);
			beg = rangebeg(range);
			end = beg+rangelen(range);
			new = listref(args, 3);
			strput(str, beg, new);
			printf("put %x %x\n", beg, end);
			return;
		}

		if(arg == "map")
			return map;

		printf("bad argument to dispatch!\n");
	}
	return mkas(dispatch);
}

dom = mkdom(c32le, mkzas(1024));

p = (char*){dom}0;
p[0] = 'h';
p[1] = 'e';
p[2] = 'l';
p[3] = 'l';

printf("%s\n", p);
