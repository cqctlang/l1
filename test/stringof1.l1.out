entry: var params, 8 locals, 0 temps
mkzas: 1 params, 3 locals, 0 temps
dispatch: var params, 7 locals, 0 temps
put 0 1
put 1 2
put 2 3
put 3 4
(null)
